전남대 BE_나제법_2주차 과제(3단계)

nove1080 opened the pull request:
## 코드 작성하면서 어려웠던 점
## 코드 리뷰 시, 멘토님이 중점적으로 리뷰해주셨으면 하는 부분
### 1. `ArgumentResolver`는 무슨 계층에 해당하나요?
`ArguementResolver`에서 Service를 참조할지 Repository에 참조할지 고민했었는데
우선 제가 만든 Service 클래스가 모두 DTO를 받고 DTO를 반환하는 형식으로 특정 Controller에만 핏하게 작성 되어있어 일단 Repository를 직접 사용하도록 만들었습니다.
어떤 방법이 더 있고 뭐가 더 권장되는지 이야기 해주실 수 있나요?

### 2. `WishProduct` 구성이 적절한가요?
```sql
#schema.sql
CREATE TABLE wish_product (
    wish_product_id BIGINT AUTO_INCREMENT PRIMARY KEY,
    member_id BIGINT NOT NULL,
    product_id BIGINT NOT NULL,
    quantity INT NOT NULL,
    FOREIGN KEY (member_id) REFERENCES member(member_id),
    FOREIGN KEY (product_id) REFERENCES product(product_id)
);
```

```java
public class WishProduct {

    private final Long id;
    private final Long memberId;
    private final Long productId;
    private Integer quantity;
}
```
회원이 위시리스트에 등록한 상품인 `WishProduct` 를 설계할 때 처음에는 첫 번째 코드와 같이 회원과 상품에 대한 Id만 가지도록 구성했었습니다. 그런데 회원과 상품의 id만 가지고 있다보니 위시리스트에 대한 API를 만들때 계속해서 회원과 상품에 대해서 구체적인 정보를 알기 위해 조회 쿼리를 보내게되어 아래와 같이 변경하였습니다.
```java
public class WishProduct {

    private final Long id;
    private final Member member;
    private final Product product;
    private Integer quantity;
}
```
```sql
SELECT wp.wish_product_id, wp.quantity,
       m.member_id, m.name AS member_name,
       p.product_id, p.name AS product_name, p.price, p.image_url 
FROM wish_product wp
JOIN member m ON wp.member_id = m.member_id 
JOIN product p ON wp.product_id = p.product_id
WHERE wp.wish_product_id = ?;
```
`WishProduct`를 조회할 때 이렇게 조회해서 한 번에 `WishProduct` 값을 모두 채워서 가져올 수 있어서 괜찮은 방법이라 생각했는데 적절한 판단이었나요? 엔티티에서 다른 엔티티를 가지고 있으면 나중에 문제가 될 수 있는 부분이 있나요?

### 3. DTO가 너무 많아진 것 같아요
추가로 궁금한 부분인데 현업에서도 API마다 모두 다른 요청,응답 DTO를 만들어서 사용하나요?
***

## 피드백 반영 후
3302b777ca66ad09697b028f4b904d7672e92003 해당 커밋부터 피드백을 적용한 작업입니다!(이렇게 알려드리면 되나요?)

### 수정한 내용
1. application.yml 민감한 정보 제거하기
2. Exception.class 예외 핸들링
3. Filter 에서 발생하는 예외 처리하기
4. 하드코딩된 값 상수로 다루기
5. ArgumentResolver에서 Service 참조하기
6. 예외 세분화하기

### 1. 필터 실행 순서를 정의하는 방법 중 어떤 것이 더 적절한가요?

> https://basalt-shirt-962.notion.site/SpringBoot-Filter-899f4ce3d4ea44d18805e18f0960a99a?pvs=74

좀 더 자세한 설명을 남겨둔 링크입니다!

1. `@Order` 어노테이션으로 지정하는 방식
8. `FilterRegistrationBean` 으로 직접 필터를 빈으로 등록하면서 순서를 정해주는 방식

위 두 가지 옵션이 존재하던데 제가 생각했을 때는 `@Order` 어노테이션으로 지정하는 방식은 등록된 필터가 많이 생겨난다면 필터 간의 실행 순서를 파악하기 한눈에 파악하기 힘들 것이라 생각하여 `FilterRegistrationBean` 을 사용하여 하나의 설정 파일에서 필터의 실행 순서를 모두 지정하는 방식을 선택하였습니다.

그런데 또 다르게 생각해보니 **정의된 필터들의 순서를 파악할 일이 있을지도 의문입니다.** 오히려 `@Order` 어노테이션 만으로 충분히 순서를 지정할 수 있고 필터 간의 실행 순서를 확인할 일이 잦을까? 있기는 할까? 필터의 실행 순서를 한 번 정해두면 다시 확인할 일은 잘 없을 것 같아 굳이 설정파일(`WebConfig`)을 더 복잡하게 만든건 아닐까? 라는 생각이 듭니다.

여기에 대해 멘토님의 의견이 궁금합니다!

### 2. `MemberDetails`와 `LoginMemberArgumentResolver`의 패키지 위치가 적절한가요?
#### 1. `LoginMemberArgumentResolver` 를 web 패키지에 둔 이유
ArgumentResolver를 현재 Controller의 파라미터를 바인딩 시켜주는 역할을 담당하니 Controller가 동작과 밀접한 관계가 있는 web 패키지에 선언하였습니다.
#### 2. `MemberDetails` 를 web 패키지에 둔 이유
`MemberDetails`역시 Controller의 파라미터로만 선언되고 있어 web 패키지에 두었습니다.

nove1080 committed:
refactor: Item의 필드를 protected로 수정

private -> protected

nove1080 committed:
feat: Item.Builder id 초기화 메서드 추가

nove1080 committed:
feat: ProductUpdateParam 추가

nove1080 committed:
feat: UpdateProductRequest 추가

nove1080 committed:
feat: UpdateProductResponse 추가

nove1080 committed:
feat: Product update 메서드 추가

nove1080 committed:
docs: ProductConverter 주석 수정

nove1080 committed:
feat: ProductRepository update 메서드 추가

nove1080 committed:
feat: ProductRepository existsById 추가

nove1080 committed:
feat: ProductService 상품 수정 기능 구현

nove1080 committed:
feat: ProductController 상품 수정 API 구현

nove1080 committed:
feat: ProductRepository 상품 수정 메서드 반환값 변경

void -> Product

nove1080 committed:
refactor: ProductRepository 상품 제거 로직 수정

nove1080 committed:
feat: ProductController 상품 추가 API 응답 코드 변경

200 -> 201

nove1080 committed:
feat: index.html 추가

nove1080 committed:
feat: ProductViewController 구현

nove1080 committed:
feat: ProductService searchProduct() 추가

nove1080 committed:
refactor: ProductController의 이름 변경 및 패키지 위치 변경

이름: ProductController -> ProductApiController
패키지: controller -> controller.api

nove1080 committed:
feat: admin.html 추가

관리자 페이지 구현

nove1080 committed:
feat: script.js

관리자 페이지에 필요한 JS 코드 파일

nove1080 committed:
feat: add-product-form.html 추가

nove1080 committed:
feat: edit-product-form.html 추가

nove1080 committed:
feat: CreateProductFormDto 추가

nove1080 committed:
feat: UpdateProductFormDto 추가

nove1080 committed:
feat: ProductToUpdateFormDtoConverter 추가

nove1080 committed:
refactor: ProductViewController Model에 entity 대신 dto 적용

nove1080 committed:
rename: ProductConverter -> CreateRequestToProductConverter

역할이 뚜렷하게 드러나도록 클래스 이름 수정

nove1080 committed:
feat: CreateRequestToProductConverter converter 인터페이스 구현

nove1080 committed:
refactor: ProductService 컨버터 리팩토링으로 인한 수정

변수 이름 및 호출 메서드명 수정

nove1080 committed:
feat: ProductToUpdateFormDtoConverter 파라미터 @NonNull 적용

nove1080 committed:
fix: ProductViewController @RequestMapping 변경

/admin -> /view

nove1080 committed:
fix: script.js 리다이렉트 주소 변경

/admin -> /view/admin

nove1080 committed:
fix: index.html 관리자 페이지 주소 변경

/admin -> /view/admin

nove1080 committed:
fix: admin.html 관리자 상품 관리 페이지 주소 변경

ProductViewController의 @RequestMapping 주소 변경으로 인한 수정

nove1080 committed:
rename: application.properties -> application.yml

nove1080 committed:
refactor: Item의 id 타입 UUID -> Long 변경

nove1080 committed:
refactor: CreateProductResponse의 Id UUID -> Long 변경

nove1080 committed:
refactor: UpdateProductResponse Id UUID -> Long 변경

nove1080 committed:
refactor: UpdateProductFormDto Id UUID -> Long 변경

nove1080 committed:
refactor: ProductRepository Id UUID -> Long 변경

nove1080 committed:
refactor: ProductViewController Id값 타입 UUID -> Long 변경

nove1080 committed:
refactor: ProductService Id값 타입 UUID -> Long 변경

nove1080 committed:
feat: schema.sql 추가

nove1080 committed:
feat: data.sql 추가

nove1080 committed:
feat: ProductRepository JdbcTemplate 추가

메모리 대신 H2DB 적용을 위함

nove1080 committed:
feat: StringToUrlConverter 추가

nove1080 committed:
chore: application.yml H2 DB 연결 코드 추가

nove1080 committed:
refactor: ProductRepository 메모리 -> 데이터베이스에 상품 저장

nove1080 committed:
refactor: ProductRepository 메모리 -> 데이터베이스에 전체 상품 조회

nove1080 committed:
refactor: ProductRepository 메모리 -> 데이터베이스에 단일 상품 조회

nove1080 committed:
refactor: ProductRepository 메모리 -> 데이터베이스에 상품 삭제

nove1080 committed:
feat: StringToUrlConverter 구현

nove1080 committed:
refactor: ProductRepository DB를 사용하도록 수정완료

nove1080 committed:
refactor: ProductApiController 파라미터로 받는 productId 타입 변경

String -> Long

nove1080 committed:
refactor: ProductRepository 사용하지 않는 import문 삭제

nove1080 committed:
refactor: Product update 메서드 삭제

DB를 사용함에 따라 Repository 에서 제품 수정을 담당하여 사용하지 않는 메서드

nove1080 committed:
refactor: ProductRepository ProductRowMapper 메서드로 분리

nove1080 committed:
refactor: Item name 필드 삭제

nove1080 committed:
rename: Item -> BaseEntity

역할을 잘 나타낼 수 있도록 이름을 수정

nove1080 committed:
refactor: Product name 필드 추가

상속받고 있는 클래스(BaseEntity)에서 name 필드를 제거함에 따른 추가

nove1080 committed:
refactor: dto를 불변 객체로 변경

nove1080 committed:
feat: CreateProductRequest Entity 변환 메서드 추가

nove1080 committed:
refactor: UpdateProductFormDto from -> fromEntity 메서드 네이밍 변경

nove1080 committed:
refactor: ProductService 변환 로직을 dto에 전가함에 따른 코드 수정

CreateRequestToProductConverter 사용 대신 dto가 직접 Entity로의 변환 책임을 가진다

nove1080 committed:
remove: CreateRequestToProductConverter

dto에서 변환 작업을 수행하도록 변경되어 사용하지 않는 클래스

nove1080 committed:
refactor: ProductViewController converter 대신 dto가 직접 변환되도록 코드를 수정

nove1080 committed:
remove: ProductToUpdateFormDtoConverter

dto에서 변환 작업을 수행하도록 변경되어 사용하지 않는 클래스

nove1080 committed:
refactor: StringToUrlConverter 정적 유틸리티 클래스로 변경

nove1080 committed:
refactor: StringToUrlConverter 정적 유틸리티 클래스로 변경에 따른 수정

nove1080 committed:
feat: ReadProductResponse 추가

domain과 view의 분리를 위함

nove1080 committed:
refactor: ReadAllProductsResponse Product 대신 ReadProductResponse 적용

nove1080 committed:
refactor: ProductService 상품 조회 응답에 Product 대신 ReadProductResponse 적용

nove1080 committed:
refactor: ProductViewController Product 대신 ReadProductResponse 적용

nove1080 committed:
refactor: Product 불변 객체로 선언

nove1080 committed:
fix: ProductRepository save 후 id가 null로 반환되는 문제 해결

nove1080 committed:
refactor: ProductViewController API path 수정

nove1080 committed:
refactor: script.js 리다이렉트 주소 수정

nove1080 committed:
refactor: admin.html 링크 주소 변경

nove1080 committed:
refactor: index.html 링크 주소 변경

nove1080 committed:
refactor: ProductService 상품 조회의 반환을 domain -> dto로 수정

nove1080 committed:
refactor: ProductRepository update 파라미터를 dto에서 domain으로 수정

ProductUpdateParam -> Product

nove1080 committed:
refactor: ProductService 불필요한 dto 삭제

ProductUpdateParam -> Product

nove1080 committed:
remove: ProductUpdateParam 삭제

사용되지 않음

nove1080 committed:
refactor: ProductApiController 리소스 수정 요청을 patch -> put 변경

nove1080 committed:
refactor: ProductViewController dto 타입 변경에 따른 수정

UpdateProductFromDto -> ReadProductResponse 사용

nove1080 committed:
refactor: script.js 수정 요청 시 patch 대신 put 으로 변경

nove1080 committed:
docs: README.md Step2 0단계 요구사항 반영

nove1080 committed:
docs: README.md Step2 1단계 요구사항 반영

nove1080 committed:
chore: build.gradle spring-boot-starter-validation 의존성 추가

nove1080 committed:
feat: RegexUtils 추가

nove1080 committed:
test: RegexUtilsTest 추가

nove1080 committed:
feat: SpecialCharacterValidator 추가

nove1080 committed:
feat: @SpecialCharacter 추가

nove1080 committed:
rename: RegexUtils -> StringUtils

nove1080 committed:
fix: StringUtils 특수문자 판별 메서드 공백 입력 케이스 처리

공백 입력의 경우 허용하지 않는 특수문자로 처리하는 문제 해결

nove1080 committed:
feat: StringUtils containsAnySubstring() 추가

nove1080 committed:
rename: CreateProductFormDto -> CreateProductForm

nove1080 committed:
rename: UpdateProductFormDto -> UpdateProductForm

nove1080 committed:
feat: application.yml message.basename=errors 추가

nove1080 committed:
feat: errors.properties 추가

nove1080 committed:
test: StringUtilsTest StringUtils.containsAnySubstring() 테스트 추가

nove1080 committed:
feat: ErrorResponse 추가

nove1080 committed:
feat: KakaoApprovalValidator 추가

nove1080 committed:
feat: @RequiredKakaoApproval 추가

nove1080 committed:
move: @SpecialCharacter validation/annotation -> validation/constraints

nove1080 committed:
move: SpecialCharacterValidator validation/annotation -> validation/constraints

nove1080 committed:
feat: CreateProductRequest BeanValidation 적용

nove1080 committed:
feat: UpdateProductRequest BeanValidation 적용

nove1080 committed:
feat: ProductApiController @Validated 적용

nove1080 committed:
feat: ProductExceptionHandler 추가

nove1080 committed:
refactor: ProductViewController import 코드 수정

nove1080 committed:
move: ErrorResponse validation -> web/dto/response

nove1080 committed:
refactor: ProductExceptionHandler import 문 수정

nove1080 committed:
style: ProductApiController 불필요한 공백 삭제

nove1080 committed:
feat: KakaoApprovalValidator 검증 시 입력값에 toLowerCase 적용

nove1080 committed:
style: errors.properties 공백 추가

nove1080 committed:
feat: errors.properties 사용하지 않는 메시지 삭제

nove1080 committed:
docs: README.md Step2 2단계 요구사항 반영

nove1080 committed:
refactor: request,response dto 패키지 이동

dto가 늘어남에 따른 관리를 위해 하위 패키지 생성

nove1080 committed:
feat: schema.sql Member 테이블 정의 추가

nove1080 committed:
feat: data.sql Member 레코드 삽입 쿼리 추가

nove1080 committed:
feat: Email 타입 추가

nove1080 committed:
feat: Password 타입 추가

nove1080 committed:
feat: Member 추가

nove1080 committed:
feat: MemberRepository 추가

nove1080 committed:
feat: MemberService 추가

nove1080 committed:
feat: @Password 추가

nove1080 committed:
feat: PasswordValidator 구현

nove1080 committed:
refactor: ProductService import문 수정

dto 패키지 세분화에 따른 경로 수정

nove1080 committed:
refactor: ProductViewController import문 수정

dto 패키지 세분화에 따른 경로 수정

nove1080 committed:
refactor: ProductApiController import문 수정

dto 패키지 세분화에 따른 경로 수정

nove1080 committed:
feat: CreateMemberRequest 추가

nove1080 committed:
feat: CreateMemberResponse 추가

nove1080 committed:
fix: ProductApiController 상품 생성 후 location 반환 관련 오타 수정

nove1080 committed:
feat: MemberApiController 추가

nove1080 committed:
chore: build.gradle jwt 의존성 추가

nove1080 committed:
feat: application.yml jwt.expiration 추가

nove1080 committed:
refactor: Email 필드 이름 수정 emailValue -> value

nove1080 committed:
refactor: Password 필드 이름 수정 passwordValue -> value

nove1080 committed:
feat: LoginRequest 추가

nove1080 committed:
feat: LoginResponse 추가

nove1080 committed:
feat: ReadMemberResponse 추가

nove1080 committed:
fix: MemberRepository 회원 저장 시 파라미터 바인딩 오류 해결

nove1080 committed:
feat: MemberRepository findByEmailAndPassword 추가

nove1080 committed:
feat: MemberService login() 추가

nove1080 committed:
feat: Token 추가

nove1080 committed:
feat: JwtProvider 추가

nove1080 committed:
refactor: MemberService 필드명 수정

nove1080 committed:
refactor: MemberApiController login 추가

nove1080 committed:
rename: ProductExceptionHandler -> ApiExceptionHandler

nove1080 committed:
style: errors.properties 파일 끝에 공백 추가

nove1080 committed:
feat: InvalidCredentialsException 추가

nove1080 committed:
feat: MemberRepository findByEmailAndPassword 예외 핸들링 추가

nove1080 committed:
feat: MemberService login 실패 시 예외 클래스 변경

nove1080 committed:
feat: ApiExceptionHandler InvalidCredentialsException 핸들링 추가

nove1080 committed:
refactor: MemberApiController readMember 삭제

nove1080 committed:
refactor: MemberApiController api path 수정

nove1080 committed:
refactor: MemberApiController api path 수정

nove1080 committed:
feat: LoginRequest email 필드 @Email 적용

nove1080 committed:
feat: JwtProvider 토큰 내용 수정

subject에 회원 아이디, claim에 회원 이메일 정보 추가

nove1080 committed:
refactor: MemberService 토큰 생성 요청 시 파라미터 수정

nove1080 committed:
docs: README.md Step2 3단계 요구사항 반영

nove1080 committed:
style: Member 불필요한 줄바꿈 삭제

nove1080 committed:
feat: application.yml logging level 설정

nove1080 committed:
feat: BaseEntity protected 생성자 추가

nove1080 committed:
feat: Member public 생성자 추가

nove1080 committed:
feat: @LoginMember 구현

nove1080 committed:
feat: JwtProvider 구현

nove1080 committed:
feat: JwtResolver 구현

nove1080 committed:
refactor: Token 클래스 패키지 위치 조정

nove1080 committed:
feat: AuthenticationFilter 추가

nove1080 committed:
refactor: LoginResponse import 경로 수정

nove1080 committed:
refactor: MemberService import 경로 수정

nove1080 committed:
feat: MemberApiController wishlist API 미완성본 추가

nove1080 committed:
feat: LoginMemberArgumentResolver 추가

nove1080 committed:
feat: WebConfig 추가

nove1080 committed:
feat: ApiExceptionHandler 전역 핸들러로 설정

nove1080 committed:
Merge branch 'nove1080' into step3

nove1080 committed:
feat: MemberProduct 추가

nove1080 committed:
feat: MemberProductRepository save() 구현

nove1080 committed:
feat: schema.sql wish_product 스키마 생성

nove1080 committed:
feat: data.sql wish_product 데이터 생성

nove1080 committed:
feat: WishProduct 추가

nove1080 committed:
remove: MemberProduct, MemberProductRepository 제거

nove1080 committed:
feat: WishProductRepository 추가

nove1080 committed:
feat: WishProductRowMapper 추가

nove1080 committed:
feat: WishProductService 추가

nove1080 committed:
feat: ReadWishProductResponse 추가

nove1080 committed:
feat: ReadAllWishProductsResponse 추가

nove1080 committed:
feat: CreateWishProductRequest 추가

nove1080 committed:
feat: CreateWishProductResponse 추가

nove1080 committed:
feat: MemberRowMapper 추가

nove1080 committed:
feat: ProductRowMapper 추가

nove1080 committed:
refactor: ProductRepository ProductRowMapper 적용 및 findById 수정

nove1080 committed:
feat: ProductApiController 위시 상품 등록 API 구현

nove1080 committed:
refactor: MemberRepository MemberRowMapper 적용 및 조회 코드 수정

nove1080 committed:
feat: MemberApiController 위시 상품 조회&삭제 API 구현

nove1080 committed:
Merge remote-tracking branch 'origin/step3' into step3

nove1080 committed:
fix: 충돌 해결

nove1080 committed:
feat: WishProduct updateQuantity() 추가

nove1080 committed:
feat: WishProductRepository findById() 추가

nove1080 committed:
feat: WishProductService updateWishProduct() 추가

nove1080 committed:
feat: UpdateWishProductRequest 추가

nove1080 committed:
feat: UpdateWishProductResponse 추가

nove1080 committed:
feat: CreateWishProductRequest productId @NotNull 추가

nove1080 committed:
feat: MemberApiController updateWishProduct API 추가

nove1080 committed:
feat: ApiExceptionHandler NoSuchElementException 핸들러 추가

nove1080 committed:
fix: 충돌 해결

Hongdonggeon commented on src/main/java/gift/authentication/filter/AuthenticationFilter.java:
```diff
@@ -0,0 +1,62 @@
+import org.springframework.web.filter.OncePerRequestFilter;
+
+@Component
+public class AuthenticationFilter extends OncePerRequestFilter {
```
filter를 이용하셨네요!  인터셉터를 이용하지 않고 filter로 선택하신 이유가 있을까요?
또한 OncePerRequestFilter를 이용하신 이유가 있을까요?

nove1080 replied:
> filter를 이용하셨네요! 인터셉터를 이용하지 않고 filter로 선택하신 이유가 있을까요? 

사용자 인증에 대해서 식별하는 작업을 filter로 많이 해둔걸 본 적이 있어서 생각없이 구현한 것 같습니다.
그래서 동건님의 코멘트를 보고 `filter`와 `interceptor`에 대해서 좀 찾아보았습니다.

다음은 `jakarta.servlet.Filter`의 주석의 일부입니다.
> Examples that have been identified for this design are 
> 1) Authentication Filters

다음과 같이 인증에 사용하기에 적합하다고 나와있고 사용자 요청에 대한 프로그램 실행 흐름을 보니 실제로 Interceptor 보다 앞에서 동작하기 때문에 인증되지 않은 사용자가 굳이 더 뒷단에서 걸러질 필요는 없다고 생각해보았습니다. 

![image](https://github.com/kakao-tech-campus-2nd-step2/spring-gift-wishlist/assets/89574219/9519eb9b-3685-4d91-b31a-9bb91bdf9005)

그런데 제가 생각했을 때는 이런 이유만으로는 interceptor 대신 filter로 인증을 처리하자고 다른 사람을 설득하기에는 명분이 좀 부족한 것 같아서 둘의 차이에 대해 좀 더 공부를 해야할 것 같습니다

> 또한 OncePerRequestFilter를 이용하신 이유가 있을까요?

이것도 습관적으로 OncePerRequestFilter 를 구현해서 만들었습니다.
스프링 MVC 구조에 대해 먼저 공부하고나면 interceptor와 filter의 동작방식의 차이와 filter의 구현체들에 대해서 공부해볼 수 있을 것 같습니다.

p.s. 지금보니 여지껏 좀 생각없이 개발을 했던 것 같네요 ㅜㅜ 앞으로는 무언가를 선택하는 상황이 있을 때 장단을 비교해보고 다른 사람을 설득할 수 있는 수준의 근거를 가지고 생각하고 개발하겠습니다 항상 좋은 고민을 해볼 수 있는 질문을 남겨주셔서 감사합니다:)

Hongdonggeon replied:
항상 제법님은 좋은 고민을 하시는 것 같아요 👍 
사실 Filter와 Interceptor의 역할은 거의 비슷하지만, 제가 아는대로 Filter와 인터셉터의 차이에 대해 말씀드릴게요
큰 차이는 필터는 자바스펙이고 인터셉터는 스프링 스펙이에요 
필스프링 어플리케이션을 셋팅하면 기본적으로 등록되는 필터와 인터셉터들이 있는데요
기본적으로 등록되는 필터와 인터셉터들의 성격을 보면, 필터는 스프링에 종속되지 않은 웹에 관련된 필터들이예요 (인코딩 관련, http 메서드들 서블릿에서 지원되도록 하는 등등 ) 
인터셉터는 (modelAttribute 바인딩하거나 더 스프링에 관련된 기술)
스프링에 종속되지 않은 웹에 관련된 기능은 필터로 구현하는게 좋을 것 같고, 더 스프링에 종속되거나 스프링에 포함된 기술을 연관해 사용하고 싶으면 인터셉터로 구현하는게 좋다고 생각해요

### Filter
- 자바 표준 스펙 
- ServletRequest,ServletResponse를 필터 체이닝 중간에 새로운 객체로 바꿀 수 있다 
- 필터에서 예외가 발생하면 `@ControllerAdvice`에서 처리하지 못한다. (스프링이 아니기 때문) 

### Interceptor
- 스프링이 제공하는 스펙
- ServletRequest,ServletResponse를 필터 체이닝 중간에 새로운 객체로 바꿀 수 없음
- 인터셉터에서 예외가 발생하면 `@ControllerAdvice`에서 처리 가능 (스프링이기 때문)

Hongdonggeon commented on src/main/java/gift/authentication/resolver/LoginMemberArgumentResolver.java:
```diff
@@ -0,0 +1,58 @@
+    public Object resolveArgument(MethodParameter parameter, ModelAndViewContainer mavContainer,
+        NativeWebRequest webRequest, WebDataBinderFactory binderFactory) throws Exception {
+
+        String authorization = webRequest.getHeader("Authorization");
```
이런 헤더명은 상수로 관리해도 좋을 것 같네요

Hongdonggeon commented on src/main/java/gift/authentication/token/Token.java:
```diff
@@ -0,0 +1,18 @@
+
+public class Token {
+
+    private String value;
```
final 변수로 선언해도 좋을 것 같아요

Hongdonggeon commented on src/main/java/gift/domain/vo/Email.java:
```diff
@@ -0,0 +1,18 @@
+        this.value = value;
+    }
+
+    public static Email from(String email) {
```
정적팩토리 메서드패턴을 선호하시네요! 정적 팩토리 메서드 패턴을 선호하신 이유가 궁금해요

nove1080 replied:
이펙티브 자바라는 책에서 본 내용인데 무작정 public 생성자를 제공하는 습관을 돌아보라고 하여서 정적 팩토리 메서드를 사용해보았습니다. 아직까지 몸으로 느낀 장점은 생성자에 이름을 부여할 수 있다는 장점 밖에 체감하지 못하고 있습니다

Hongdonggeon replied:
흠 또 어떤 장점이 있을까요..
객체를 여러번 만들지 않아도 되는 경우에는 미리 생성해서 정적 팩토리 메서드로 생성해놓은 객체를 반환할 수 도 있을 것 같고, 반환 타입을 인터페이스로 선언해두고 조건에 따라 하위 구현체를 반환할수도 있겠네요

정리하자면 정적 팩토리 메서드는 말 그대로 메서드니까 생성자보다 더 커스텀하게 만들 수 있다는 장점이 있는 것 같네요~

Hongdonggeon commented on src/main/java/gift/repository/MemberRepository.java:
```diff
@@ -0,0 +1,62 @@
+            Member findMember = jdbcTemplate.queryForObject(sql, getMemberRowMapper(), email.getValue(), password.getValue());
+            return Optional.ofNullable(findMember);
+        } catch (EmptyResultDataAccessException e) {
+            return Optional.empty();
```
예외 처리후에 emtpy로 반환하는 이유가 있을까요?
optional null로 반환해도 되지 않을까해서요!

nove1080 replied:
Optional null로 반환한다는게 무슨 말인지 잘 모르겠습니다 ㅜㅜ

Hongdonggeon replied:
아 그냥 예외처리를 하지않고 Optiona.ofNUllable로 반환해도 되지 않나하는 이야기였어요
그래서 이 repository 메서드를 사용하는 곳에서 null인 경우에 예외처리를 해도 되지 않나 하는 말이였습니다

nove1080 replied:
```java
Member findMember = jdbcTemplate.queryForObject(sql, getMemberRowMapper(), email.getValue(), password.getValue());
return Optional.ofNullable(findMember);
```
처음에는 저도 코드를 위와 같이 예외처리를 하지 않고 `Optional.ofNullable()` 로 반환하도록 구성했었습니다.
그리고 DB에 없는 회원을 조회하는 테스트를 해보았는데요 

없는 회원을 조회한 경우 `Optional.EMPTY`가 반환되고 정상적으로 프로그램이 실행될 것이라고 기대했는데
`Optional.EMPTY`가 반환되지 않고 `EmptyResultDataAccessException` 이 발생하더라구요

https://stackoverflow.com/questions/10606229/jdbctemplate-query-for-string-emptyresultdataaccessexception-incorrect-result

위 글을 읽어보니 `JdbcTemplate.queryForObject()` 반드시 하나의 결과 행이 포함되지 않으면 예외를 발생시키게 된다고 하더라구요 `jdbcTemplate.query()`를 이용하거나 `try-catch`로 감싸는 방법 중 저는 후자를 선택하였습니다.

그러면 메서드 반환타입에 `Optional` 없앨 수도 있겠네요!

Hongdonggeon replied:
넵 그렇네요 예외가 발생하지 않는다는 다는건 조회한 데이터가 존재하는거니까 Optional을 없앨 수 있겠네요

Hongdonggeon commented on src/main/resources/application.yml:
```diff
@@ -15,4 +18,12 @@ spring:
+      data-locations: classpath:sql/data.sql
+
+jwt:
+  secretkey: 994e69b6df494d3e5f48fdb32693d5c772cc35a4d4f67cce4323fe256b6caf31b666d21db4b9f974935c7900bce11947edf8e266ef87aaa176db591b220860ff
```
시크릿키가 github에 이렇게 공개적으로 노출되면 보안에 취약할 것 같은데요. 다른 yml 파일을 만들어서 노출되지 않도록 관리하는게 좋을 것 같아요

Hongdonggeon commented on src/main/java/gift/web/dto/response/wishproduct/UpdateWishProductResponse.java:
```diff
@@ -0,0 +1,53 @@
+import gift.domain.Product;
+import gift.domain.WishProduct;
+
+public class UpdateWishProductResponse {
```
ReadwishProductResponse와 객체 구조가 동일하네요! 
WisthProductResponse와 같이 네이밍 변경하고 공통적으로 사용해도 되지 않을까요?

Hongdonggeon reviewed (changes requested):
제법님 미션 구현하느라 수고 많으셨습니다! 
요구사항에 잘 충족해주셨고, 추가로 코드도 깔끔해서 리뷰하기 편했습니다 👍
코멘트들 반영하면서 궁금한 것 있으시면 편하게 DM 주세요!

Hongdonggeon commented:
- ArgumentResolver는 무슨 계층에 해당하나요?
> ArgumentResolver를 컨트롤러에서 사용되니, 컨트롤러 계층이 가장 잘 어울리지 않을까 생각해요
argumentResolver에서 서비스를 참조할지, repository를 참조할지에 대해서도 물어보셨는데 정말 좋은 고민이라고 생각합니다! 
정답은 없겠지만 argumentResolver가 컨트롤러 계층이 가깝다고 생각이 들어서 service를 참조하는게 더 적절하지 않을까 싶어요

- wishProduct 설계
> 저는 적절한 판단이었다고 생각해요! 
객체에서 모두 id만 들고 있다면 제법님이 말씀하신 것처럼 조회쿼리가 빈번해질거고 
객체에서 너무 모든 객체를 가지고 있다면 추후에 객체 연관관계를 잘 고려해야 될 거예요 
지금 설계에서 약간의 코멘트를 드리자면,
wishProduct의 product 객체가 사용되고 있는데, member 객체는 사용되고 있지 않고 연관관계의 역할만 해주고 있잖아요
이런 경우에는 wishtProduct 가 memberId 정보만 들고 있어도 됐었을 것 같다는 생각도 들긴하는 것 같아요! (하나의 의견입니다 고민한번 해주시면 좋을 것 같아요)


- 추가로 궁금한 부분인데 현업에서도 API마다 모두 다른 요청,응답 DTO를 만들어서 사용하나요?
> 이건 케이스마다 다를 것 같아요
모든 api마다 모두 다른 요청 응답 dto를 무조건적으로 만들어야되는 건 아니구요! api가 각각 독립적이라면 따로 dto를 관리하겠지만, 이 미션의
updateProductResponse와 readProductResponse 모두 product 도메인의 정보를 읽어오는거고 객체 구조가 동일하기 때문에 이런건 공통적으로 사용할 수 있을 것 같아요

nove1080 committed:
remove: UpdateProductForm.java

사용하지 않음

nove1080 committed:
refactor: Token value 필드 final 추가

nove1080 committed:
feat: JsonUtils 추가

nove1080 committed:
refactor: ErrorResponse 필드 이름 변경 field -> reason

nove1080 committed:
refactor: AuthenticationFilter 로그 제거

nove1080 committed:
feat: ApiExceptionHandler.java Exception.class 에러 핸들링 추가

nove1080 committed:
feat: AuthenticationExceptionHandlerFilter.java 추가

nove1080 committed:
feat: WebConfig.java 커스텀 필터 빈 등록

- AuthenticationExceptionHandlerFilter
- AuthenticationFilter
실행 우선 순위를 제어하기 위함

Hongdonggeon reviewed (approved):
제법님 안녕하세요! 
미션 수행하시느라 수고 많으셨습니다 :) 
이미 잘하시지만, 이번 미션하면서 정말 많이 성장하신 것 같아요 👍
열심히 학습하고 고민하신 성과가 아닌 가 싶습니다!
저도 제법님 리뷰하면서 많이 배우고 있습니다 
코멘트 반영하신 것 잘 확인했고 approve 할게요 
지금 컨플릭 나는 파일들이 많아서 제법님이 컨플리 해결해주시면 제가 머지해드리겠습니다!

nove1080 committed:
chore: .gitignore application-secret.yml 추가

nove1080 committed:
refactor: application.yml 민감한 정보 제거

- DB 아이디, 패스워드, JWT 시크릿 키 제거

nove1080 committed:
feat: application-secret.yml 추가

nove1080 committed:
fixed untracked files

nove1080 committed:
refactor: JsonUtils.java 추상 클래스로 선언

nove1080 commented:
> 제법님 안녕하세요! 미션 수행하시느라 수고 많으셨습니다 :) 이미 잘하시지만, 이번 미션하면서 정말 많이 성장하신 것 같아요 👍 열심히 학습하고 고민하신 성과가 아닌 가 싶습니다! 저도 제법님 리뷰하면서 많이 배우고 있습니다 코멘트 반영하신 것 잘 확인했고 approve 할게요 지금 컨플릭 나는 파일들이 많아서 제법님이 컨플리 해결해주시면 제가 머지해드리겠습니다!

아직 다 리뷰를 다 반영하지 못한 것 같기도 하고 코드를 좀 더 리팩토링 해보고 싶어서 내일 안으로 충돌까지 해결해서 다시 리뷰요청을 드려도 괜찮을까요?

nove1080 committed:
feat: Category 추가

nove1080 committed:
feat: ErrorCode 추가

nove1080 committed:
refactor: ErrorResponse ErrorCode 사용하도록 수정

nove1080 committed:
feat: IncorrectEmailException.java 추가

nove1080 committed:
feat: IncorrectPasswordException.java 추가

nove1080 committed:
refactor: InvalidCredentialsException.java 예외 메시지 ErrorCode 사용

nove1080 committed:
refactor: AuthenticationExceptionHandlerFilter.java ErrorCode 사용

nove1080 committed:
refactor: LoginMemberArgumentResolver.java 인증 헤더명 상수로 추출

nove1080 committed:
refactor: AuthenticationFilter.java 매직 리터럴 상수로 추출

nove1080 committed:
feat: Password.java matches() 추가

nove1080 committed:
feat: Member.java matchPassword() 추가

nove1080 committed:
feat: MemberRepository findByEmail() 추가

nove1080 committed:
refactor: MemberService.java login 중 예외 발생 세분화

nove1080 committed:
refactor: 커스텀 검증 어노테이션 ErrorCode 사용하도록 수정

nove1080 committed:
refactor: AuthenticationFilter.java 매직 넘버 상수로 추출

nove1080 committed:
refactor: ErrorCode.java NOT_FOUND 예외 메시지 변경

nove1080 committed:
feat: MemberDetails.java 추가

nove1080 committed:
feat: MemberDetailsService.java 추가

nove1080 committed:
refactor: LoginMemberArgumentResolver.java 도메인 계층 의존성 제거

nove1080 committed:
refactor: MemberApiController.java 도메인 의존성 제거

nove1080 committed:
refactor: ProductApiController.java 도메인 의존성 제거

nove1080 committed:
refactor: 패키지 위치 조정

nove1080 committed:
feat: ApiExceptionHandler.java 핸들러 추가

- IncorrectEmailException
- IncorrectPasswordException
- NoResourceFoundException

nove1080 committed:
remove: errors.properties

nove1080 committed:
feat: application.yml message.basename 제거

nove1080 committed:
Merge branch 'nove1080' into step3

nove1080 committed:
remove: ProductExceptionHandler.java

nove1080 committed:
fix: 중복 빈 등록 문제 해결

nove1080 committed:
remove: errors.properties

Hongdonggeon commented:
> > 제법님 안녕하세요! 미션 수행하시느라 수고 많으셨습니다 :) 이미 잘하시지만, 이번 미션하면서 정말 많이 성장하신 것 같아요 👍 열심히 학습하고 고민하신 성과가 아닌 가 싶습니다! 저도 제법님 리뷰하면서 많이 배우고 있습니다 코멘트 반영하신 것 잘 확인했고 approve 할게요 지금 컨플릭 나는 파일들이 많아서 제법님이 컨플리 해결해주시면 제가 머지해드리겠습니다!
> 
> 아직 다 리뷰를 다 반영하지 못한 것 같기도 하고 코드를 좀 더 리팩토링 해보고 싶어서 내일 안으로 충돌까지 해결해서 다시 리뷰요청을 드려도 괜찮을까요?

리뷰 요청 하신 것 같아서 머지할게요! 
3주차 과제 진행하면서도 리팩토링 계속하시면서 질문 언제든지 주세요! 
머지 할게요 ~
